_posts/2020-06-28-openfaas-loki-into.md

```diff
@@ -0,0 +1,235 @@
+
+Loki, when installed in Kubernetes, will automatically create a search index for each of the lables on the Pods it sees. Important, this means the Loki query for the logs of an OpenFaaS function looks like `{faas_function="<name>"}` because OpenFaaS always adds the function name as the label `faas_function`.  This [link](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bfaas_function%3D%5C%22cows%5C%22%7D%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D) will show the logs for the Cows function we installed and invoked.
+
+### Installing `openfaas-loki`
+
+To combine OpenFaaS and Loki, we need to install `openfaas-loki`:
+
+```bash
+arkade install openfaas-loki
+```
+
+We then need to configure the OpenFaaS installation, we can do this with `arkade` using the `--set` flag
+
+```bash
+arkade install openfaas --basic-auth=false --set gateway.logsProviderURL=http://openfaas-loki.default:9191/
+```
+
+## Testing the logs
+
+### Recreate the port-forward
+Earlier, while we were installing and verifying the system, we ran this command
+
+```bash
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Unfortunately, that connection is no longer valid, because we updated the OpenFaaS installation to configure the logs provider.  We need to stop the current `port-forward` command and then start a new one:
+
+```bash
+pkill kubectl
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+### Invoke the function and check the logs
+Now we can use the `faas-cli` to invoke and see the logs, this time powered by Loki.
+
+```bash
+faas-cli invoke cows <<< ""
+             ___   ___
+             \  \  \  \
+   (___)   ___\__\__\__\__
+   (o o)   |  O O O O O O|
+  --\ /----+-------------+-------/
+  |  O                          /
+   \                           /
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~s~~~~~~~~~~~
+         Steerage
+```
+and
+
+```bash
+faas-cli logs cows
+2020-06-27T18:24:09Z 2020-06-27T18:24:09.633228Z stderr F 2020/06/27 18:24:09 Forking fprocess.
+2020-06-27T18:24:09Z 2020-06-27T18:24:09.8101492Z stderr F 2020/06/27 18:24:09 Wrote 279 Bytes - Duration: 0.176972 seconds
+```
+
+
+To compare with the Kubernetes powered logs
+
+```bash
+2020-06-27T16:37:17Z 2020/06/27 16:37:17 Forking fprocess.
+2020-06-27T16:37:17Z 2020/06/27 16:37:17 Wrote 119 Bytes - Duration: 0.130287 seconds
+```
+
+We see a slightly different format that is coming from the way Loki scrapes the logs. Primarily
+
+```
+2020-06-27T18:24:09.633228Z stderr F
+```
+
+This shows the timestamp for when the log was collected and that it came from `stderr` instead of `stdout`.
+
+We can also demonstrate that the logs are now independently stored from the function instance by scaling the function to 0 and still retrieving the same logs
```

Comment:
I like this point. This is a huge benefit and I would like to see this point at the beginning of the post as I think it's not quite clear enough.

Something like this in the introduction: With the default configuration, we read logs for a function from the Kubelet, which means that if a Pod is scaled down or removed, that we also lose its history. By using a log aggregator like Loki, we can query it even after the Pod has been removed.

Reply:
Re-reading the intro I can see that you have covered this. Thanks 👍